apis/v1alpha2/grpcroute_types.go

```diff
@@ -0,0 +1,552 @@
+	// Support: Core (Exact with service and method specified)
+	//
+	// Support: Implementation-specific (Exact with method specified but no service specified)
+	//
+	// Support: Implementation-specific (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	//
+	// At least one of Service and Method MUST be a non-empty string.
+	// +optional
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:Pattern=`^[^\/]*$`
+	Service *string `json:"service,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	//
+	// At least one of Service and Method MUST be a non-empty string.
+	// +optional
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:Pattern=`^[^\/]*$`
+	Method *string `json:"method,omitempty"`
+}
+
+// MethodMatchType specifies the semantics of how gRPC methods and services are compared.
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
+//
+// Exact paths MUST be syntactically valid:
+//
+// - Must not contain `/` character
+//
+// +kubebuilder:validation:Enum=Exact;RegularExpression
+type GRPCMethodMatchType string
+
+const (
+	// Matches the method or service exactly and with case sensitivity.
+	GRPCMethodMatchExact GRPCMethodMatchType = "Exact"
+
+	// Matches if the method or service matches the given regular expression with
+	// case sensitivity.
+	//
+	// Since `"RegularExpression"` has implementation-specific conformance,
+	// implementations can support POSIX, PCRE, RE2 or any other regular expression
+	// dialect.
+	// Please read the implementation's documentation to determine the supported
+	// dialect.
+	GRPCMethodMatchRegularExpression GRPCMethodMatchType = "RegularExpression"
+)
+
+// GRPCHeaderMatch describes how to select a gRPC route by matching gRPC request
+// headers.
+type GRPCHeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name is the name of the gRPC Header to be matched.
+	//
+	// If multiple entries specify equivalent header names, only the first
+	// entry with an equivalent name MUST be considered for a match. Subsequent
+	// entries with an equivalent header name MUST be ignored. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered
+	// equivalent.
+	Name HeaderName `json:"name"`
+
+	// Value is the value of the gRPC Header to be matched.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=4096
+	Value string `json:"value"`
+}
+
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=256
+// +kubebuilder:validation:Pattern=`^[A-Za-z0-9!#$%&'*+\-.^_\x60|~]+$`
+type HeaderName string
```

Comment:
Why did we have it to alias string here?

Reply:
This was actually not sharing types with HTTP, updated to make that more clear in #1576. The type aliases are intended to ensure the same concept has consistent validation throughout the API.